GEOMESA-780 Kafka Data Store Replay

mmatz-ccri opened the pull request:
* Added replay ability to Kafka Data Store.
* GEOMESA-799 Kafka consumer with ability to start from arbitrary offset

Signed-off-by: Michael Matz <mmatz@ccri.com>

jnh5y commented on geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/filter/package.scala:
```diff
@@ -307,33 +306,7 @@ package object filter {
-  def checkOrderUnsafe(one: Expression, two: Expression): PropertyLiteral =
-    checkOrder(one, two)
-        .getOrElse(throw new RuntimeException("Expressions did not contain valid property and literal"))
+  val checkOrderUnsafe = org.locationtech.geomesa.filter.checkOrderUnsafe _
```
Do we need to forward these functions or can just relocate them?  (Which I'm totally in favor of.)

mmatz-ccri replied:
Sure, I'll relocate them.  Or, maybe relocate the whole class to org.locationtech.geomesa.filter (same as it was before zcurve)?  The only problem with that is FilterPackageObjectTest depends on TestData in geomesa-accumulo-datastore/src/test.

jnh5y commented on geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/AttributeIdxStrategy.scala:
```diff
@@ -35,6 +35,7 @@ import org.locationtech.geomesa.accumulo.index.QueryPlanners.JoinFunction
 import org.locationtech.geomesa.accumulo.index.Strategy._
 import org.locationtech.geomesa.accumulo.iterators._
 import org.locationtech.geomesa.features.SerializationType.SerializationType
+import org.locationtech.geomesa.filter.PropertyLiteral
```
Is this import used?

mmatz-ccri replied:
Yes, it was covered by import org.locationtech.geomesa.accumulo.filter._ until I moved PropertyLiteral.

jnh5y commented on geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/PropertyLiteral.scala:
```diff
@@ -6,7 +6,7 @@
  */
 
-package org.locationtech.geomesa.accumulo.filter
+package org.locationtech.geomesa.filter
```
Should this move between sub-modules?  (I haven't checked the full path quite yet.)

mmatz-ccri replied:
I moved it back where it was before zcurve because geomesa-kafka-datastore needs it.  There's nothing accumulo specific about it.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/README.md:
```diff
@@ -0,0 +1,51 @@
+    
+###Data Producers
+
+Data *producers* should build a SimpleFeatureType object and pass it to `prepareForLive` along with a  zookeeper path that 
```
I wonder if 'producers' should use a separate prepareToProduce method.  If the implementation can call between the prepareToProduce and prepareForLive, that's fine.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/README.md:
```diff
@@ -0,0 +1,51 @@
+    
+    ...
+    
+    val zkPath = "/a/unique/zkpath"
```
What happens if the zkPath is not-unique/reused?

mmatz-ccri replied:
The zkPath is not required to be unique.  It's purpose is to provide disambigation if needed.  There may be multiple SFTs under the same zkPath.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/README.md:
```diff
@@ -0,0 +1,51 @@
+along with a `ReplayConfig`) to create the derived schema via a call to `DataStore.prepareForReplay()`.
+
+    val datastore = ...
+    val typename = "FrOut"
```
Let's use a more general sounding name.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/README.md:
```diff
@@ -0,0 +1,51 @@
+    datastore.createSchema(KafkaDataStoreHelper.prepareForReplay(sft,rc))
+
+It should be noted that the new schema will have a different type name and an additional attribute in the schema (time) 
+from the original SimpleFeatureType
```
(Pedantic) Sentences end with periods

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/pom.xml:
```diff
@@ -70,6 +79,13 @@
+            <groupId>net.sf.jopt-simple</groupId>
+            <artifactId>jopt-simple</artifactId>
+            <version>3.2</version>
+            <scope>compile</scope>
```
1.  Do we have a IP CQ in for this dependency?  If not, I'll need some details.
2.  Let's try and manage the version in the root pom.

mmatz-ccri replied:
Its was a transitive dependency.  I'm now using it directly but could elminiate it with a little work.

mmatz-ccri replied:
I've moved the version management to the parent where there is already some Kafka related dependency management.  Going forwared should all dependency management be added / moved to the root?

Additionally, I noticed that geomesa-kafka was using a different version of Guava than that managed by the root.  I removed the override.

elahrvivaz replied:
Should double check that - storm might require a different version of guava.

mmatz-ccri replied:
Storm is using a 3rd version of guava.  What a mess!  I put the guava version back as it was however it seems to me that using different version of guava within different modules of geomesa is not good.  geomesa-kafka-geoserver-plugin is shading and relocating it so external requirements should not affect what version we use but I'll leave this cleanup for another day.

mmatz-ccri replied:
I've marked jopt-simple as provided.  Currently kafka is not a provided dependency.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/GeoMessage.scala:
```diff
@@ -0,0 +1,184 @@
+  */
+case class CreateOrUpdate(override val timestamp: Instant, feature: SimpleFeature) extends GeoMessage {
+
+  val id = feature.getID
```
I'd recommend making this a def.

mmatz-ccri replied:
It was only used in one place so I removed it.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/GeoMessage.scala:
```diff
@@ -0,0 +1,184 @@
+  *
+  * The following encoding is used:
+  *
+  * Key: version (1 byte) type (1 byte) timstamp (8 bytes)
```
Let's not tramp this Tim fellow.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/GeoMessage.scala:
```diff
@@ -0,0 +1,184 @@
+      case c: CreateOrUpdate => createOrUpdateType
+      case d: Delete => deleteType
+      case x: Clear => clearType
+      case u => throw new IllegalArgumentException(s"Invalid message: '$u'")
```
I'm tempted to suggest using a 'sealed' trait here for GeoMessage.

mmatz-ccri replied:
It already is a sealed trait.  The last case is mainly to handle null although it should not be null...

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/GeoMessage.scala:
```diff
@@ -0,0 +1,184 @@
+    case d: Delete =>
+      encodeDeleteMessage(d)
+    case x: Clear =>
+      encodeClearMessage(x)
```
For clarity, I'd suggest removing the 'import GeoMessageEncoder._' line and calling encodeClearMessage and encodeDeleteMessage prefixed by GeoMessageEncoder.  

I searched for these functions and I found a few with the same name and different signatures.  It seems like there's potential for confusion.

mmatz-ccri replied:
I did some refactoring to clean this up a bit.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/GeoMessage.scala:
```diff
@@ -0,0 +1,184 @@
+  val serializer: SimpleFeatureDeserializer = new KryoFeatureSerializer(schema, SerializationOptions.withUserData)
+
+  def decode(key: Array[Byte], msg: Array[Byte]): GeoMessage = {
+    val MsgKey(version, msgType, ts) = decodeKey(key)
```
Is MsgKey only used to encode/extract for this line?  If so, a Tuple3 should suffice.

mmatz-ccri replied:
It is used elsewhere.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/GeoMessage.scala:
```diff
@@ -0,0 +1,184 @@
+    if (key == null) {
+      throw new IllegalArgumentException("Invalid null key.")
+    }
+    if (key.length != 10) {
```
I'd suggest checking the version and throwing an exception based on that.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStore.scala:
```diff
@@ -135,42 +59,55 @@ object KafkaDataStoreFactoryParams {
 
-  override def createNewDataStore(params: ju.Map[String, Serializable]): DataStore = ???
+  override def createNewDataStore(params: ju.Map[String, Serializable]): DataStore =
+    throw new UnsupportedOperationException
```
Hmm... did the compiler make you add this?  I imagine the override would have taken care of the exception declaration.  

I think it'll be fine either way.

elahrvivaz replied:
I guess it's just explicitly unsupported vs not implemented yet.

mmatz-ccri replied:
Exactly, it not supported.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreHelper.scala:
```diff
@@ -0,0 +1,103 @@
+
+    val builder = FeatureUtils.builder(sft)
+    val preparedSft = builder.buildFeatureType()
+    insertTopic(preparedSft,buildTopicName(zkPath,sft))
```
(Picky) Space after the comma.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreHelper.scala:
```diff
@@ -0,0 +1,103 @@
+    builder.setName(buildReplayTypeName(sft.getTypeName))
+    ReplayTimeHelper.addReplayTimeAttribute(builder)
+
+    // assumes topic has been prepared, need to check though...
```
If there's a follow-on ticket, feel free to create it and reference it in the comment.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreHelper.scala:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  /** modifies the passed SimpleFeature type, adding the topic string to user data */
+  def insertTopic(sft: SimpleFeatureType, topic: String) : SimpleFeatureType = {
```
This function can return 'Unit' since it is mutating state.  (All uses I observed follow that pattern.

jnh5y replied:
This is optional.

jnh5y commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreHelper.scala:
```diff
@@ -0,0 +1,103 @@
+  def isPreparedForLive(sft: SimpleFeatureType): Boolean =
+    sft.getUserData.containsKey(TopicKey) && !sft.getUserData.containsKey(ReplayConfigKey)
+
+  def isPreparedForReplay(sft: SimpleFeatureType): Boolean = sft.getUserData.containsKey(ReplayConfigKey)
```
Is the TopicKey ever set for 'replay'?

mmatz-ccri replied:
Yes, it is.  The SFT must be preparedForLive before it can be preparedForReplay.

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/README.md:
```diff
@@ -0,0 +1,51 @@
+    
+    
+    KafkaDataStoreHelper.prepareForLive(sft: SimpleFeatureType, zkPath: String) : SimpleFeatureType
+    KafkaDataStoreHelper.prepareForReplay(sft: SimpleFeatureType, rc: ReplayConfig) : SimpleFeatureType
```
Could we rename these to something like 'configureForReplay' - or if the original SFT is not modified, something like 'createReplaySft' ?

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaConsumerFeatureSource.scala:
```diff
@@ -16,124 +16,65 @@
-
-  val cb = CacheBuilder.newBuilder()
+  override def getBoundsInternal(query: Query) =
+    ReferencedEnvelope.create(new Envelope(-180, 180, -90, 90), DefaultGeographicCRS.WGS84)
```
Should use the shared constant for 4326 here - i think it's in simple feature types. It will use the full crs if available.

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreHelper.scala:
```diff
@@ -0,0 +1,103 @@
+import org.opengis.feature.simple.SimpleFeatureType
+import org.locationtech.geomesa.utils.geotools.Conversions.RichSimpleFeatureType
+
+object KafkaDataStoreHelper {
```
I think I would move this all into the KafkaDataStore object - we don't normally have 'helper' classes in scala as the companion object takes that place. @jnh5y thoughts?

mmatz-ccri replied:
The KafkaDataStoreHelper is responsible for managing the userdata in SFT required by KafkaDataStore. Its enough of its own thing that I'd like to keep it in its own file.  However, I'm not opposed to renaming it if you can suggest something better.

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreSchemaManager.scala:
```diff
@@ -0,0 +1,183 @@
+
+  //
+  // todo:  consistent exception handling across all methods
+  // todo   - catch and rethrow a known/consistent exception with helpful msg
```
should make tickets or delete the comments.

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreSchemaManager.scala:
```diff
@@ -0,0 +1,183 @@
+    kfc.replayConfig.foreach(r => createZkNode(getReplayConfigPath(typeName), ReplayConfig.encode(r)))
+
+    //create the Kafka topic
+    if(!AdminUtils.topicExists(zkClient, kfc.topic))
```
use { }

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreSchemaManager.scala:
```diff
@@ -0,0 +1,183 @@
+    }.getOrElse(None)
+  }
+
+  override def getNames: util.List[Name] = zkClient.getChildren(zkPath).asScala.map(
```
[picky] formatting - maybe move everything after equals down a line so it fits on one. otherwise, split on . like:
foo.bar
  .baz

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreSchemaManager.scala:
```diff
@@ -0,0 +1,183 @@
+    zkClient.deleteRecursive(getSchemaPath(typeName))
+
+    // todo: what about other remote caches - do we set up a kafka listener looking for deletes?
+    // todo: what about kafka - who has the responsibility to clean the kafka topic up?
```
tickets?

mmatz-ccri replied:
I replaced the TODOs with a reference to this ticket:
https://geomesa.atlassian.net/browse/GEOMESA-818

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreSchemaManager.scala:
```diff
@@ -0,0 +1,183 @@
+
+  private val zkClient = {
+    // zkStringSerializer is required - otherwise topics won't be created correctly
+    val ret = new ZkClient(zookeepers, Int.MaxValue, Int.MaxValue, ZKStringSerializer)
```
we should really be closing this zkClient at some point... maybe override the 'destroy' method in datastore? (might be named something else, but i think there is a lifecycle method)

mmatz-ccri replied:
added close in dispose()

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreSchemaManager.scala:
```diff
@@ -0,0 +1,183 @@
+  *                                  ``KafkaDataStoreHelper.prepareForLive``
+  */
+@throws[IllegalArgumentException]
+private[kafka] case class KafkaFeatureConfig(sft: SimpleFeatureType) extends AnyRef {
```
hmm, is there some significance to a case class extending AnyRef? Just curious.

mmatz-ccri replied:
Nope, it was probably an AnyVal at one point and changed rather than removed.

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaDataStoreSchemaManager.scala:
```diff
@@ -0,0 +1,183 @@
+  val replayConfig: Option[ReplayConfig] = KafkaDataStoreHelper.extractReplayConfig(sft)
+
+  override def toString: String =
+    s"KafkaSimpleFeatureType: typeName=${sft.getTypeName}; topic=$topic; replayConfig=$replayConfig"
```
this should probably be 'KafkaFeatureConfig:' ?

mmatz-ccri replied:
good catch

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessageEncoding.scala:
```diff
@@ -0,0 +1,78 @@
+  *
+  * @param schema the [[SimpleFeatureType]]; required to serialize [[CreateOrUpdate]] messages
+  */
+class KafkaGeoMessageEncoder(schema: SimpleFeatureType) extends GeoMessageEncoder(schema) {
```
Is it overkill to have the separate GeoMessageEncoder at this point? If we only have 1 implementation might be clearer to just have the logic here.

mmatz-ccri replied:
probably

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessageFormatter.scala:
```diff
@@ -0,0 +1,176 @@
+  *
+  * @see KafkaDataStoreLogViewer for an alternative to kafka-console-consumer
+  */
+class KafkaGeoMessageFormatter extends MessageFormatter {
```
nice - a formatter for the accumulo shell would be sweet too :)

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessageFormatter.scala:
```diff
@@ -0,0 +1,176 @@
+
+    if(!props.containsKey(sftSpecKey)) {
+      throw new IllegalArgumentException(s"Property '$sftSpecKey' is required.")
+    }
```
[minor] scala has a 'require' function that does this in a slightly cleaner manner.

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaGeoMessageFormatter.scala:
```diff
@@ -0,0 +1,176 @@
+  * To run, first copy the geomesa-kafka-geoserver-plugin.jar to $KAFKA_HOME/libs.  Then create a copy of
+  * $KAFKA_HOME/bin/kafka-console-consumer.sh called "kafka-ds-log-viewer" and in the copy replace the
+  * classname in the exec command at the end of the script with
+  * org.locationtech.geomesa.kafka.KafkaDataStoreLogViewer.
```
If these aren't in the readme's already, please copy them there.

mmatz-ccri replied:
done

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/LiveKafkaConsumerFeatureSource.scala:
```diff
@@ -0,0 +1,163 @@
+
+    val cb = CacheBuilder.newBuilder().ticker(ticker)
+
+    expirationPeriod.map { ep =>
```
prefer foreach over map to indicate the side effect

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/ReplayKafkaConsumerFeatureSource.scala:
```diff
@@ -0,0 +1,419 @@
+    reader.getOrElse(new EmptyFeatureReader[SimpleFeatureType, SimpleFeature](replaySFT))
+  }
+
+  private[kafka] def snapshot(time: Option[Long]): Option[ReplaySnapshotFeatureCache] = {
```
private[package] seems an odd combination... might want it to be private[this] ?

mmatz-ccri replied:
I want to be able to unit test the function directly but not make it part of the public api.

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/ReplayKafkaConsumerFeatureSource.scala:
```diff
@@ -0,0 +1,419 @@
+
+      if (index < 0) {
+        // no message found at sought time
+        index = -index - 1
```
is this equivalent to index = 0? that would be clearer...

mmatz-ccri replied:
It's not equivalent.  The binary search returns the "index of the search key, if it is contained in the array otherwise, <tt>(-(<i>insertion point</i>) - 1)</tt>".  This calculation is to get the correct inseition point.  I added a comment to clarify.

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/ReplayKafkaConsumerFeatureSource.scala:
```diff
@@ -0,0 +1,419 @@
+
+      // there may be multiple messages at the same time so there may be messages before ``index``
+      // that are at the given ``time``
+      while (index > 0 && messages(index - 1).timestamp.getMillis <= time) index -= 1
```
{ }

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/ReplayKafkaConsumerFeatureSource.scala:
```diff
@@ -0,0 +1,419 @@
+    })
+
+    // required: there is only 1 partition;  validate??
+    val stream = kafkaConsumer.createMessageStreams(1, offsetRequest).head
```
this is how it's always worked, right? only a single partition.

mmatz-ccri replied:
That what I thought but there is a configuration parameters for the number of partitions.  Why was that parameter added?  Could we support muliple partitions if messages are partitioned by the simple feature ID?  If so, replay would need some more work to handle.  Should the number of partitions parameter be removed or should I create a ticket for reviewing this another time?

elahrvivaz replied:
It should be fine to have a single consumer against multiple partitions I think, you just don't get any parallelism so it's kind of pointless. I would create a ticket to revisit the issue.

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/ReplayKafkaConsumerFeatureSource.scala:
```diff
@@ -0,0 +1,419 @@
+  require(attrIndex >= 0, s"Invalid SFT.  The $AttributeName attribute is missing.")
+
+  /** Copy the given ``sf`` and add a value for the replay time attribute. */
+  def addReplayTime(sf: SimpleFeature): SimpleFeature = {
```
i would rename this to make clear that a new feature is being created.

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/ReplayKafkaConsumerFeatureSource.scala:
```diff
@@ -0,0 +1,419 @@
+
+  private def processMessages: (Quadtree, mutable.Map[String, FeatureHolder]) = {
+    val features = new mutable.HashMap[String, FeatureHolder]()
+    val qt = new Quadtree
```
does this need to be a synchronized qt?

mmatz-ccri replied:
No, each query to a replay consumer feature source results in the creation of a new ReplaySnapshotFeatureCache used only to answer that single query and then its thrown away.

elahrvivaz commented on geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/ReplayKafkaConsumerFeatureSource.scala:
```diff
@@ -0,0 +1,419 @@
+    val childSplits = children.flatMap(c => split(c))
+
+    if (childSplits.size != children.size) {
+      // one or more children are invalid
```
does this logic work for an OR? couldn't you still get a valid timestamp from one side?

mmatz-ccri replied:
I could handle each side of the OR and then union the results.  But, you might end up with multiple different versions of the same simple feature at different points in time.  There might be a use case for this but there is also an issues with the different versions having the same ID but not being equal.  Currently we don't need to support this, as far as I know.

elahrvivaz commented on geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/InfiniteIterator.scala:
```diff
@@ -0,0 +1,46 @@
+
+import scala.collection.Iterator.empty
+
+object InfiniteIterator {
```
woah

elahrvivaz commented on geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/SimpleFeatureTypes.scala:
```diff
@@ -35,6 +38,10 @@ import scala.util.parsing.combinator.JavaTokenParsers
 
+  type FR = FeatureReader[SimpleFeatureType, SimpleFeature]
+  type DFR = DelegateFeatureReader[SimpleFeatureType, SimpleFeature]
+  type DFI = DelegateFeatureIterator[SimpleFeature]
```
hmm, this doesn't seem like a great place for these... maybe move to a package object somewhere?

mmatz-ccri replied:
moved

elahrvivaz commented:
looks good to me. i didn't review the kafka consumer stuff, since i wrote it (so it's obviously perfect ;)

mmatz-ccri commented:
Thanks elahrvivaz.  You might want to take a quick look at the consumer stuff since I changed it a bit.  I had an issue with the binary search (NPE referencying a lazy val) so I rewrote it a little to simplify the logic and added a unit tested it.

Afterglow375 commented on geomesa-kafka/geomesa-kafka-datastore/README.md:
```diff
@@ -0,0 +1,73 @@
+###Data Producers
+
+Data *producers* should build a SimpleFeatureType object and pass it to `prepareForLive` along with a zookeeper path that 
+must be unique for a given KafkaDataStore.  The Consumer should pass the returned SimpleFeature type to the call to
```
Typo, consumer doesn't have anything to do with Datastore.createSchema.

AndrewAnnex commented:
Merged in: 5595bdaeac207ddfb7903246f3a97bc45b6e0c88
